docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+# Consumption tracking
+
+# Goals
+
+This proposal is made with two mostly but not entirely overlapping goals:
+
+* Collect info that is needed for consumption-based billing
+* Cross-check AWS bills
+
+
+# Metrics
```

Comment:
I'd like to propose to meter cross-AZ traffic on SK level to, it costs us money. In future, we need to understand what to do with this.

Reply:
Not sure what that means in practical terms. All the WAL traffic that goes in and out of the safekeeper is metered by that `last_lsn` which is also reported to compute and pageserver. So regardless of where we meter it value would be the same. Also neither safekeeper not pageserver have any information about whether counterpart is in the same AZ or not.

But we can deduce how much times WAL traffic is crossing AZ boundaries by tacking into account deploy topology. Basically there are two variants:

![Untitled - Frame 2](https://user-images.githubusercontent.com/284219/203315055-00b8fbff-920c-47d5-a83a-239ddcc866b8.jpg)

In the first setup WAL will cross AZ boundaries two times, and in a second setup it will cross it three times. We don't have too much control now over which safekeeper is selected (it is done automatically) but we can report the current source safekeeper in the event so that we can keep track of it.

There is also a question in close placement of proxy and compute and compute and pageserver. Ideally all the pipeline in healthy situation should look like that:

![Untitled - Frame 3](https://user-images.githubusercontent.com/284219/203317100-5a3871ab-842f-46e8-ba00-cf8f688a7e4d.jpg)

Right now we are more or less randomly throw that boxes in AZ's, so on average we cross AZ lots of times.

But let's discuss that separately -- both how to detect cross-az traffic and how to better colocate services.

Reply:
agreed to discuss it separately